File: ognibuild/fix_build.py
#!/usr/bin/python3
# Copyright (C) 2020 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

from functools import partial
import logging
from typing import List, Tuple, Callable, Optional, TypeVar

from buildlog_consultant import Problem
from buildlog_consultant.common import (
    find_build_failure_description,
    MissingCommand,
)

from . import DetailedFailure, UnidentifiedError
from .session import Session, run_with_tee


# Number of attempts to fix a build before giving up.
DEFAULT_LIMIT = 200


class FixerLimitReached(Exception):
    """The maximum number of fixes has been reached."""


class BuildFixer:
    """Build fixer."""

    def can_fix(self, problem: Problem):
        raise NotImplementedError(self.can_fix)

    def _fix(self, problem: Problem, phase: Tuple[str, ...]):
        raise NotImplementedError(self._fix)

    def fix(self, problem: Problem, phase: Tuple[str, ...]):
        if not self.can_fix(problem):
            return None
        return self._fix(problem, phase)


def run_detecting_problems(
        session: Session, args: List[str], check_success=None,
        quiet=False, **kwargs) -> List[str]:
    error: Optional[Problem]
    if not quiet:
        logging.info('Running %r', args)
    if check_success is None:
        def check_success(retcode, contents):
            return (retcode == 0)
    try:
        retcode, contents = run_with_tee(session, args, **kwargs)
    except FileNotFoundError:
        error = MissingCommand(args[0])
        retcode = 1
    else:
        if check_success(retcode, contents):
            return contents
        lines = "".join(contents).splitlines(False)
        match, error = find_build_failure_description(lines)
        if error is None:
            if match:
                logging.warning("Build failed with unidentified error:")
                logging.warning("%s", match.line.rstrip("\n"))
            else:
                logging.warning(
                    "Build failed and unable to find cause. Giving up.")
            raise UnidentifiedError(retcode, args, lines, secondary=match)
    raise DetailedFailure(retcode, args, error)


T = TypeVar('T')


def iterate_with_build_fixers(
        fixers: List[BuildFixer],
        cb: Callable[[], T], limit=DEFAULT_LIMIT) -> T:
    """Call cb() until there are no more DetailedFailures we can fix.

    Args:
      fixers: List of fixers to use to resolve issues
      cb: Callable to run the build
      limit: Maximum number of fixing attempts before giving up
    """
    attempts = 0
    fixed_errors = []
    while True:
        to_resolve = []
        try:
            return cb()
        except DetailedFailure as e:
            to_resolve.append(e)
        while to_resolve:
            f = to_resolve.pop(-1)
            logging.info("Identified error: %r", f.error)
            if f.error in fixed_errors:
                logging.warning(
                    "Failed to resolve error %r, it persisted. Giving up.",
                    f.error
                )
                raise f
            attempts += 1
            if limit is not None and limit <= attempts:
                raise FixerLimitReached(limit)
            try:
                resolved = resolve_error(f.error, None, fixers=fixers)
            except DetailedFailure as n:
                logging.info("New error %r while resolving %r", n, f)
                if n in to_resolve:
                    raise
                to_resolve.append(f)
                to_resolve.append(n)
            else:
                if not resolved:
                    logging.warning(
                        "Failed to find resolution for error %r. Giving up.",
                        f.error
                    )
                    raise f
                fixed_errors.append(f.error)


def run_with_build_fixers(
     fixers: Optional[List[BuildFixer]], session: Session, args: List[str],
     quiet: bool = False, **kwargs
) -> List[str]:
    if fixers is None:
        fixers = []
    return iterate_with_build_fixers(
        fixers,
        partial(run_detecting_problems, session, args, quiet=quiet, **kwargs))


def resolve_error(error, phase, fixers) -> bool:
    relevant_fixers = []
    for fixer in fixers:
        if fixer.can_fix(error):
            relevant_fixers.append(fixer)
    if not relevant_fixers:
        logging.warning("No fixer found for %r", error)
        return False
    for fixer in relevant_fixers:
        logging.info("Attempting to use fixer %s to address %r", fixer, error)
        made_changes = fixer.fix(error, phase)
        if made_changes:
            return True
    return False

File: src/fix_build.rs
use buildlog_consultant::{Match, Problem};
use log::{info, warn};
use std::fmt::{Debug, Display};

/// A fixer is a struct that can resolve a specific type of problem.
pub trait BuildFixer<O: std::error::Error>: std::fmt::Debug + std::fmt::Display {
    /// Check if this fixer can potentially resolve the given problem.
    fn can_fix(&self, problem: &dyn Problem) -> bool;

    /// Attempt to resolve the given problem.
    fn fix(&self, problem: &dyn Problem) -> Result<bool, InterimError<O>>;
}

#[derive(Debug)]
/// Intermediate error type used during build fixing.
///
/// This enum represents different kinds of errors that can occur during
/// the build process, and which may be fixable by a BuildFixer.
pub enum InterimError<O: std::error::Error> {
    /// A problem that was detected during the build, and that we can attempt to fix.
    Recognized(Box<dyn Problem>),

    /// An error that we could not identify.
    Unidentified {
        /// The return code of the failed command.
        retcode: i32,
        /// The output lines from the command.
        lines: Vec<String>,
        /// Optional secondary information about the error.
        secondary: Option<Box<dyn Match>>,
    },

    /// Another error raised specifically by the callback function that is not fixable.
    Other(O),
}

/// Error result from repeatedly running and attemptin to fix issues.
#[derive(Debug)]
pub enum IterateBuildError<O> {
    /// The limit of fixing attempts was reached.
    FixerLimitReached(usize),

    /// A problem was detected that was recognized but could not be fixed.
    Persistent(Box<dyn Problem>),

    /// An error that we could not identify.
    Unidentified {
        /// The return code of the failed command.
        retcode: i32,
        /// The output lines from the command.
        lines: Vec<String>,
        /// Optional secondary information about the error.
        secondary: Option<Box<dyn Match>>,
    },

    /// Another error raised specifically by the callback function that is not fixable.
    Other(O),
}

impl<O: Display> Display for IterateBuildError<O> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            IterateBuildError::FixerLimitReached(limit) => {
                write!(f, "Fixer limit reached: {}", limit)
            }
            IterateBuildError::Persistent(p) => {
                write!(f, "Persistent build problem: {}", p)
            }
            IterateBuildError::Unidentified {
                retcode,
                lines,
                secondary,
            } => write!(
                f,
                "Unidentified error: retcode: {}, lines: {:?}, secondary: {:?}",
                retcode, lines, secondary
            ),
            IterateBuildError::Other(e) => write!(f, "Other error: {}", e),
        }
    }
}

impl<O: std::error::Error> std::error::Error for IterateBuildError<O> {}

/// Call cb() until there are no more DetailedFailures we can fix.
///
/// # Arguments
/// * `fixers`: List of fixers to use to resolve issues
/// * `cb`: Callable to run the build
/// * `limit: Maximum number of fixing attempts before giving up
pub fn iterate_with_build_fixers<
    T,
    // The error type that the fixers can return.
    I: std::error::Error,
    // The error type that the callback function can return, and the eventual return type.
    E: From<I> + std::error::Error,
>(
    fixers: &[&dyn BuildFixer<I>],
    mut cb: impl FnMut() -> Result<T, InterimError<E>>,
    limit: Option<usize>,
) -> Result<T, IterateBuildError<E>> {
    let mut attempts = 0;
    let mut fixed_errors: std::collections::HashSet<Box<dyn Problem>> =
        std::collections::HashSet::new();
    loop {
        let mut to_resolve: Vec<Box<dyn Problem>> = vec![];

        match cb() {
            Ok(v) => return Ok(v),
            Err(InterimError::Recognized(e)) => to_resolve.push(e),
            Err(InterimError::Unidentified {
                retcode,
                lines,
                secondary,
            }) => {
                return Err(IterateBuildError::Unidentified {
                    retcode,
                    lines,
                    secondary,
                });
            }
            Err(InterimError::Other(e)) => return Err(IterateBuildError::Other(e)),
        }

        while let Some(f) = to_resolve.pop() {
            info!("Identified error: {:?}", f);
            if fixed_errors.contains(&f) {
                warn!("Failed to resolve error {:?}, it persisted. Giving up.", f);
                return Err(IterateBuildError::Persistent(f));
            }
            attempts += 1;
            if let Some(limit) = limit {
                if limit <= attempts {
                    return Err(IterateBuildError::FixerLimitReached(limit));
                }
            }
            match resolve_error(f.as_ref(), fixers) {
                Err(InterimError::Recognized(n)) => {
                    info!("New error {:?} while resolving {:?}", &n, &f);
                    if to_resolve.contains(&n) {
                        return Err(IterateBuildError::Persistent(n));
                    }
                    to_resolve.push(f);
                    to_resolve.push(n);
                }
                Err(InterimError::Unidentified {
                    retcode,
                    lines,
                    secondary,
                }) => {
                    return Err(IterateBuildError::Unidentified {
                        retcode,
                        lines,
                        secondary,
                    });
                }
                Err(InterimError::Other(e)) => return Err(IterateBuildError::Other(e.into())),
                Ok(resolved) if !resolved => {
                    warn!("Failed to find resolution for error {:?}. Giving up.", f);
                    return Err(IterateBuildError::Persistent(f));
                }
                Ok(_) => {
                    fixed_errors.insert(f);
                }
            }
        }
    }
}

/// Attempt to resolve a problem using available fixers.
///
/// # Arguments
/// * `problem` - The problem to resolve
/// * `fixers` - List of fixers to try
///
/// # Returns
/// * `Ok(true)` - If the problem was fixed
/// * `Ok(false)` - If no fixer could fix the problem
/// * `Err(InterimError)` - If fixing the problem failed
pub fn resolve_error<O: std::error::Error>(
    problem: &dyn Problem,
    fixers: &[&dyn BuildFixer<O>],
) -> Result<bool, InterimError<O>> {
    let relevant_fixers = fixers
        .iter()
        .filter(|fixer| fixer.can_fix(problem))
        .collect::<Vec<_>>();
    if relevant_fixers.is_empty() {
        warn!("No fixer found for {:?}", problem);
        return Ok(false);
    }
    for fixer in relevant_fixers {
        info!("Attempting to use fixer {} to address {:?}", fixer, problem);
        let made_changes = fixer.fix(problem)?;
        if made_changes {
            return Ok(true);
        }
    }
    Ok(false)
}

/// Run a command repeatedly, attempting to fix any problems that occur.
///
/// This function runs a command and applies fixes if it fails,
/// potentially retrying multiple times with different fixers.
///
/// # Arguments
/// * `fixers` - List of fixers to try if the command fails
/// * `limit` - Optional maximum number of fix attempts
/// * `session` - The session to run the command in
/// * `args` - The command and its arguments
/// * `quiet` - Whether to suppress output
/// * `cwd` - Optional current working directory
/// * `user` - Optional user to run as
/// * `env` - Optional environment variables
///
/// # Returns
/// * `Ok(Vec<String>)` - The output lines if successful
/// * `Err(IterateBuildError)` - If the command fails and can't be fixed
pub fn run_fixing_problems<
    // The error type that the fixers can return.
    I: std::error::Error,
    // The error type that the callback function can return.
    E: From<I> + std::error::Error + From<std::io::Error>,
>(
    fixers: &[&dyn BuildFixer<I>],
    limit: Option<usize>,
    session: &dyn crate::session::Session,
    args: &[&str],
    quiet: bool,
    cwd: Option<&std::path::Path>,
    user: Option<&str>,
    env: Option<&std::collections::HashMap<String, String>>,
) -> Result<Vec<String>, IterateBuildError<E>> {
    iterate_with_build_fixers::<Vec<String>, I, E>(
        fixers,
        || {
            crate::analyze::run_detecting_problems(
                session,
                args.to_vec(),
                None,
                quiet,
                cwd,
                user,
                env,
                None,
            )
            .map_err(|e| match e {
                crate::analyze::AnalyzedError::Detailed { retcode: _, error } => {
                    InterimError::Recognized(error)
                }
                crate::analyze::AnalyzedError::Unidentified {
                    retcode,
                    lines,
                    secondary,
                } => InterimError::Unidentified {
                    retcode,
                    lines,
                    secondary,
                },
                crate::analyze::AnalyzedError::MissingCommandError { command } => {
                    InterimError::Recognized(Box::new(
                        buildlog_consultant::problems::common::MissingCommand(command),
                    ))
                }
                crate::analyze::AnalyzedError::IoError(e) => InterimError::Other(e.into()),
            })
        },
        limit,
    )
    .map_err(|e| match e {
        IterateBuildError::Other(e) => IterateBuildError::Other(e.into()),
        e => e,
    })
}
